### PR TITLE
Add support for framedumping to OpenGL ES.

### PR DESCRIPTION
**Pull request by Sonicadvance1**
[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3426)

### CODE REVIEW

File: Source/Core/VideoCommon/AVIDump.h [hunk 1]
```diff
@@ -25,10 +25,16 @@ class AVIDump
 	static void* GetFrame();
 
 public:
+	enum DumpFormat
```
Comment (lioncash) [inline Source/Core/VideoCommon/AVIDump.h]: enum class

File: Source/Core/VideoCommon/AVIDump.cpp [hunk 2]
```diff
@@ -53,8 +53,11 @@ static void* s_stored_frame = nullptr;
 static u64 s_stored_frame_size = 0;
 static bool s_start_dumping = false;
 
-bool AVIDump::Start(HWND hWnd, int w, int h)
+bool AVIDump::Start(HWND hWnd, int w, int h, AVIDump::DumpFormat format)
```
Comment (lioncash) [inline Source/Core/VideoCommon/AVIDump.cpp]: `AVIDump::` as a prefix on DumpFormat isn't necessary

File: Source/Core/VideoCommon/AVIDump.cpp [hunk 3]
```diff
@@ -359,8 +363,13 @@ static void InitAVCodec()
 	}
 }
 
-bool AVIDump::Start(int w, int h)
+bool AVIDump::Start(int w, int h, AVIDump::DumpFormat format)
```
Comment (lioncash) [inline Source/Core/VideoCommon/AVIDump.cpp]: same here

File: Source/Core/VideoCommon/AVIDump.cpp [hunk 4]
```diff
@@ -359,8 +363,13 @@ static void InitAVCodec()
...
 {
+	if (format == AVIDump::DumpFormat::FORMAT_BGR)
```
Comment (lioncash) [inline Source/Core/VideoCommon/AVIDump.cpp]: and here

File: Source/Core/VideoBackends/D3D/Render.cpp [hunk 5]
```diff
@@ -840,7 +840,7 @@ void Renderer::SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, co
 		{
 			s_recordWidth = source_width;
 			s_recordHeight = source_height;
-			bAVIDumping = AVIDump::Start(D3D::hWnd, s_recordWidth, s_recordHeight);
+			bAVIDumping = AVIDump::Start(D3D::hWnd, s_recordWidth, s_recordHeight, AVIDump::DumpFormat::FORMAT_BGR);
```
Comment (RisingFog) [inline Source/Core/VideoBackends/D3D/Render.cpp]: Is this needed for D3D?

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 6]
```diff
@@ -1342,110 +1342,80 @@ void Renderer::SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, co
 
 	// Frame dumps are handled a little differently in Windows
 	// Frame dumping disabled entirely on GLES3
```
Comment (RisingFog) [inline Source/Core/VideoBackends/OGL/Render.cpp]: Comment can be removed

File: Source/Core/VideoCommon/AVIDump.cpp [hunk 7]
```diff
@@ -53,8 +53,11 @@ static void* s_stored_frame = nullptr;
 static u64 s_stored_frame_size = 0;
 static bool s_start_dumping = false;
 
-bool AVIDump::Start(HWND hWnd, int w, int h)
+bool AVIDump::Start(HWND hWnd, int w, int h, DumpFormat format)
 {
+	// Win32 only supports BGR format
+	(void)format;
+
```
Comment (RisingFog) [inline Source/Core/VideoCommon/AVIDump.cpp]: Doesn't this defeat the entire purpose of having format?

Comment (BhaaLseN) [inline Source/Core/VideoCommon/AVIDump.cpp]: Fair point, if you leave the Windows signature without the format type, it should be just fine. It has its own signature anyways for the HWND, so you could just drop it.
...unless some later feature (DX12?) could use it. Which might also depend on the planned replacement of VFW with ffmpeg, which might not require the HWND anymore.

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 8]
```diff
@@ -1342,110 +1342,80 @@ void Renderer::SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, co
...
 
-				bLastFrameDumped = true;
-			}
-			else
-			{
-				NOTICE_LOG(VIDEO, "Error reading framebuffer");
-			}
-		}
-		else
+		if (GLInterface->GetMode() == GLInterfaceMode::MODE_OPENGLES3)
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/Render.cpp]: Why is this conditionally? Just always export as RGBX.

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 9]
```diff
@@ -1342,110 +1342,80 @@ void Renderer::SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, co
...
-			bLastFrameDumped = false;
+			bytes_per_pixel = 4;
+			data_type = GL_RGBA;
+			dump_fmt = AVIDump::DumpFormat::FORMAT_RGBA;
 		}
-#else
-		if (SConfig::GetInstance().m_DumpFrames)
+
+		if (frame_data.empty() ||
+		    w != flipped_trc.GetWidth() || h != flipped_trc.GetHeight())
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/Render.cpp]: A no-op resizing call is very cheap. so there is no need to have a global state for w and h. Just always call resize.

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 10]
```diff
@@ -1342,110 +1342,80 @@ void Renderer::SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, co
...
-			glPixelStorei(GL_PACK_ALIGNMENT, 1);
-			glReadPixels(GetTargetRectangle().left, GetTargetRectangle().bottom, w, h, GL_BGR, GL_UNSIGNED_BYTE, &frame_data[0]);
+			w = flipped_trc.GetWidth();
+			h = flipped_trc.GetHeight();
+			frame_data.resize(bytes_per_pixel * w * h);
+		}
 
+		glPixelStorei(GL_PACK_ALIGNMENT, 1);
+		glReadPixels(flipped_trc.left, flipped_trc.bottom, w, h, data_type, GL_UNSIGNED_BYTE, &frame_data[0]);
+		if (w > 0 && h > 0)
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/Render.cpp]: So if the resolution is 0x0 - or negative, you still call glReadPixels?

File: Source/Core/VideoCommon/AVIDump.cpp [hunk 11]
```diff
@@ -53,8 +53,11 @@ static void* s_stored_frame = nullptr;
 static u64 s_stored_frame_size = 0;
 static bool s_start_dumping = false;
 
-bool AVIDump::Start(HWND hWnd, int w, int h)
+bool AVIDump::Start(HWND hWnd, int w, int h, DumpFormat format)
 {
+	// Win32 only supports BGR format
```
Comment (degasus) [inline Source/Core/VideoCommon/AVIDump.cpp]: This comment is wrong, only this lines need to be touched: https://github.com/Sonicadvance1/dolphin/blob/ES_fix_framedump/Source/Core/VideoCommon/AVIDump.cpp#L289

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 12]
```diff
@@ -1337,64 +1337,73 @@ void Renderer::SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, co
...
 	// Frame dumping disabled entirely on GLES3
-	if (GLInterface->GetMode() == GLInterfaceMode::MODE_OPENGL)
-	{
 #if defined(HAVE_LIBAV) || defined (_WIN32)
-		if (SConfig::GetInstance().m_DumpFrames)
+	if (SConfig::GetInstance().m_DumpFrames)
+	{
+		std::lock_guard<std::mutex> lk(s_criticalScreenshot);
+		int bytes_per_pixel = 3;
+		GLenum data_type = GL_BGR;
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/Render.cpp]: What is wrong with always using RGBA?

Comment (Sonicadvance1) [inline Source/Core/VideoBackends/OGL/Render.cpp]: Fear of getting yelled at like I did when previously forcing RGBA.

**Comment by degasus** [comment]
LGTM

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3426#-:-K8s9puZ7me1ltF2G227:676834233)*

**Pull request merged**